Atmos Auth Implementation (#1475)
cloudposse/atmos: website/docs/cli/commands/auth/usage.mdx

---
title: atmos auth
sidebar_label: auth
sidebar_class_name: command
description: Atmos Authentication Commands
---

import DocCardList from '@theme/DocCardList';
import Screengrab from '@site/src/components/Screengrab'
import Intro from '@site/src/components/Intro'

<Intro>
Atmos Auth provides a unified authentication system for cloud providers, supporting identity chaining and credential management. Use this section to get started and to understand the architecture.
</Intro>

## Usage

<Screengrab title="atmos auth --help" slug="atmos-auth--help" />

## Examples

```shell
# Validate configuration
atmos auth validate

# Authenticate with the default identity
atmos auth login

# Authenticate with a specific identity
atmos auth login --identity admin

# Print environment variables in JSON
atmos auth env --format json

# Execute a command with authentication context
atmos auth exec -- terraform plan

# Show current authentication status
atmos auth whoami
```

## Arguments

<dl>
  <dt>n/a</dt>
  <dd>No positional arguments. Use subcommands.</dd>

>   

Reviewer: ```suggestion
```

## Quick Start

### 1. Basic Configuration

Add authentication configuration to your `atmos.yaml`:

```yaml
auth:
  # Configure logging
  ## Optional, defaults to Atmos log level
  logs:
    level: Info

  providers:
    my-sso:
      kind: aws/iam-identity-center
      region: us-east-1
      start_url: https://mycompany.awsapps.com/start

  identities:
    admin:
      kind: aws/permission-set
      default: true
      via:
        provider: my-sso
      principal:
        name: AdminAccess
        account:
          name: "account-name"
```

Notes:

- Region is required for the GitHub OIDC provider and is validated at construction time.

### 2. Validate Configuration

```bash
atmos auth validate
```

### 3. Authenticate

```bash
# Use default identity
atmos auth login

# Use specific identity
atmos auth login --identity admin
```

### 4. Check Authentication Status

```bash
atmos auth whoami
```

### 5. Use with Terraform

```bash
# Atmos automatically handles authentication
atmos terraform plan mycomponent -s dev
```

## Authentication Concepts

### Providers

> **Providers** are the root authentication sources that obtain initial credentials:

Reviewer: ```suggestion
**Providers** are the upstream systems that Atmos Auth uses to obtain initial credentials:
```

- **AWS SSO**: `aws/iam-identity-center`
- **AWS SAML**: `aws/saml`
- **GitHub OIDC**: `github/oidc`

### Identities

> **Identities** use provider credentials to assume specific roles or permissions:

Reviewer: ```suggestion
**Identities** represent the user accounts or roles available from provider credentials, such as:
```

- **AWS Permission Set**: `aws/permission-set`
- **AWS Assume Role**: `aws/assume-role`
- **AWS User**: `aws/user`

### Identity Chaining

> You can chain identities to create complex authentication flows:

Reviewer: ```suggestion
Identity chaining (often called *role chaining*) is when one identity is used to obtain another, forming a sequence of temporary credentials.  

For example, you might:  
1. Start with an SSO login to obtain base credentials.  
2. Use those credentials to assume a cross-account role.  
3. Optionally, chain again into another role with more limited or specialized permissions.  

This allows you to:  
- Access multiple accounts or environments without long-lived keys.  
- Follow least-privilege practices by escalating only as needed.  
- Automate complex authentication flows while still relying on short-lived credentials.  
- 

You can chain identities to create complex authentication flows like this:
```

atmos auth env --format dotenv

# Execute command with authentication
atmos auth exec --identity prod-admin -- aws sts get-caller-identity
atmos auth exec -- terraform plan

# Configure AWS user credentials
atmos auth user configure
```

## Component-Level Configuration

Override authentication settings for specific components:

```yaml
# In your component configuration
components:
  terraform:
    myapp:
      auth:
        identities:
          custom-role:
            kind: aws/assume-role
            via:
              provider: my-sso
            principal:
              assume_role: arn:aws:iam::123456789012:role/MyAppRole
```

## Environment Variable Formats

### Bash Format

```bash
atmos auth env --format bash
# Output:
export AWS_ACCESS_KEY_ID="AKIA..."
export AWS_SECRET_ACCESS_KEY="..."
export AWS_SESSION_TOKEN="..."
```

### JSON Format

```bash
atmos auth env --format json
# Output:
{
  "AWS_ACCESS_KEY_ID": "AKIA...",
  "AWS_SECRET_ACCESS_KEY": "...",
  "AWS_SESSION_TOKEN": "..."
}
```

### Dotenv Format

```bash
atmos auth env --format dotenv
# Output:
AWS_ACCESS_KEY_ID=AKIA...
AWS_SECRET_ACCESS_KEY=...
AWS_SESSION_TOKEN=...
```

## Default Identity Handling

> 

Reviewer: ```suggestion

A **default identity** is the one Atmos Auth will use automatically when no specific identity is requested.  

- If you configure a single identity and mark it as `default: true`, Atmos will always use it without requiring you to pass `--identity`.  
- If multiple identities are defined, you can still mark one as default, but you’ll need to explicitly choose another when you don’t want the default.  
- If no default is set and multiple identities exist, Atmos will require you to specify which identity to use.  

This makes workflows simpler because you can avoid passing extra flags for the identity you use most often.

```

### Single Default

When you have one default identity, it's used automatically:

```yaml
identities:
  admin:
    kind: aws/permission-set
    default: true # This will be used by default
```

### Multiple Defaults

When multiple defaults exist, Atmos behavior depends on the environment:

**Interactive Mode**: Prompts you to choose

```bash
$ atmos auth whoami
? Multiple default identities found. Please choose one:
  ▸ dev-admin
    prod-admin
    staging-admin
```

### No Defaults

**Interactive Mode**: Shows all available identities

```bash
$ atmos auth whoami
? No default identity configured. Please choose an identity:
  ▸ dev-admin
    prod-admin
    staging-admin
```

## Credential Storage

Atmos securely stores credentials using your operating system's keyring:

- **macOS**: Keychain
- **Linux**: Secret Service (GNOME Keyring, KDE Wallet)
- **Windows**: Windows Credential Manager

### Credential Expiration

- Credentials are automatically refreshed when expired
- You can check expiration with `atmos auth whoami`
- Manual refresh: `atmos auth login --identity <name>`

## AWS File Management

> Atmos manages AWS credential files separately from your personal AWS configuration:

Reviewer: ```suggestion
Atmos manages AWS credential files separately from your personal AWS configuration.  

This design avoids common problems: 

1. **No risk of overwriting personal config**: Atmos never modifies `~/.aws/credentials` or `~/.aws/config`. This prevents accidental overrides of credentials you manage yourself and keeps your personal AWS setup intact.

2. **Provider isolation**: Each provider’s credentials live under `~/.aws/atmos/<provider>/`, making it safe and easy to switch between SSO, SAML, OIDC, or static identities without conflicts.

3. **Seamless SDK integration**: Under the hood, Atmos sets the standard AWS SDK environment variables (`AWS_SHARED_CREDENTIALS_FILE`, `AWS_CONFIG_FILE`) so tools that already rely on the AWS SDK (like Terraform, AWS CLI, CDK) just work without extra setup.

4. **Project convenience**: Since all configuration lives under Atmos, every project is self-contained. You don’t have to share custom AWS config steps with teammates, and you can cleanly delete all credentials or config associated with a profile at any time.

This makes authentication more secure, more portable across projects, and less error-prone in day-to-day development.
```
